File: find_intersection_LL.py
class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        lenA=0
        lenB=0
        curr=headA
        while(curr):
            lenA+=1
            curr=curr.next
        curr=headB
        while(curr):
            lenB+=1
            curr=curr.next
        while lenA>lenB:
            lenA-=1
            headA=headA.next
        while lenB>lenA:
            lenB-=1
            headB=headB.next
        while headA!=headB:
            headA=headA.next
            headB=headB.next
        return headA

#Time-Complexity: O(m+n)
#Space-Complexity: O(1)  as no extra space is used

File: FindIntersectionLL.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach


/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode(int x) {
 * val = x;
 * next = null;
 * }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int lenA = 0;
        int lenB = 0;

        ListNode tempA = headA;
        while (tempA != null) {
            lenA++;
            tempA = tempA.next;
        }

        ListNode tempB = headB;
        while (tempB != null) {
            lenB++;
            tempB = tempB.next;
        }

        tempA = headA;
        tempB = headB;

        if (lenB > lenA) {

            for (int i = 0; i < lenB - lenA; i++) {
                tempB = tempB.next;
            }

            for (int i = 0; i < lenA; i++) {
                if(tempA == tempB) {
                    return tempA;
                }
                tempB = tempB.next;
                tempA = tempA.next;
            }
            
        } else {
             for (int i = 0; i < lenA - lenB; i++) {
                tempA = tempA.next;
            }

            for (int i = 0; i < lenB; i++) {
                if(tempA == tempB) {
                    return tempA;
                }
                tempB = tempB.next;
                tempA = tempA.next;
            }
        }

        return null;
    }
}